### PR TITLE
Handle shard lose during inserting markers

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Handle shard lose during inserting markers

**Why?**
Only the shard who has the ownership should insert failover marker

**How did you test it?**
integration tests

**Potential risks**
N/A

### CODE REVIEW

File: common/persistence/cassandra/cassandraPersistence.go [hunk 1]
```diff
@@ -2925,5 +2944,34 @@ func (d *cassandraPersistence) CreateFailoverMarkerTasks(
...
+			if rangeID, ok := previous["range_id"].(int64); ok && rangeID != request.RangeID {
+				// CreateWorkflowExecution failed because rangeID was modified
+				return &p.ShardOwnershipLostError{
+					ShardID: d.shardID,
+					Msg: fmt.Sprintf("Failed to create workflow execution.  Request RangeID: %v, Actual RangeID: %v",
+						request.RangeID, rangeID),
+				}
+			}
+		}
+		// At this point we only know that the write was not applied.
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/persistence/cassandra/cassandraPersistence.go]: Updated

File: service/history/shard/context.go [hunk 2]
```diff
@@ -1195,7 +1195,12 @@ func (s *contextImpl) InsertFailoverMarkers(
 
 	var err error
 	for attempt := 0; attempt < conditionalRetryCount; attempt++ {
-		err = s.executionManager.CreateFailoverMarkerTasks(&persistence.CreateFailoverMarkersRequest{Markers: markers})
+		err = s.executionManager.CreateFailoverMarkerTasks(
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/history/shard/context.go:1198]: Updated

File: service/history/shard/context.go [hunk 3]
```diff
@@ -1194,10 +1194,25 @@ func (s *contextImpl) InsertFailoverMarkers(
...
+				RangeID: s.getRangeID(),
+				Markers: markers,
+			},
+		)
+		switch err.(type) {
+		case nil:
 			break
+		case *persistence.ShardOwnershipLostError:
+			// do not retry on ShardOwnershipLostError
+			break
```
Comment (yycptt) [inline service/history/shard/context.go:1209]: `s.closeShard()` to trigger a shard close?

Comment (yux0) [reply service/history/shard/context.go:1210]: good point. Updated

**Pull request merged**